### ISSUE TITLE
MauiApp not Disposed in maui-blazor

### ISSUE THREAD

**Issue by malciin** [dotnet/maui#20350]
### Description

I have noticed that all of my `Singleton` services are not properly disposed in my app. Upon further investigation I've noticed thats probably (I'm newbie in maui) because neither of:

https://github.com/dotnet/maui/blob/a6c99972618fe6d042e8f066ec479cd1b41b2a67/src/Core/src/Hosting/MauiApp.cs#L38

https://github.com/dotnet/maui/blob/a6c99972618fe6d042e8f066ec479cd1b41b2a67/src/Core/src/Hosting/MauiApp.cs#L46

was called.

### Steps to Reproduce

1. Create singleton service that implements IDisposable
2. Set breakpoint in dispose code.
3. Close window by clicking X

Results: Dispose never gets called thus breakpoint never occur.

Reproduction repo: https://github.com/malciin/maui-blazor-dispose-issue/commit/e031693c7a23f1c4b830ce093f581c50be127d2d

### Link to public reproduction project repository

https://github.com/malciin/maui-blazor-dispose-issue/commit/e031693c7a23f1c4b830ce093f581c50be127d2d

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

Windows, I was *not* able test on other platforms

**Comment by drasticactions** [comment]
https://github.com/drasticactions/MauiRepros/tree/main/DisposeTest

Doing this in a regular MAUI app causes the same behavior to occur. MauiApp is disposable, but nothing I can find so far shows that anything invokes it to be disposed of upon app close. If anything, MauiApp is [created](https://github.com/dotnet/maui/blob/96288c1cff1c6399f42e3e24ea6f5b54557a19c0/src/Core/src/Platform/Windows/MauiWinUIApplication.cs#L32-L36) and its references are used to build `MauiContext`, but `MauiApp` itself isn't attached to anything once it's been called. 

Now, that seems kinda weird to me (Why make the object just to create a new thing off of it and throw it away) but as for Dispose itself, I don't think this is a bug. To me, that makes sense: Your app is closing. The resources you were using are going to be released anyway. Calling disposed doesn't make sense there. And if you were doing logic within your app in its Disposed methods... I would rethink that approach.

If you were using MAUI in an [embedded app](https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/native-embedding?view=net-maui-8.0) case, where you create MauiApp yourself and control its existence, being able to dispose it at will makes sense, as you may only need it for a specific part of your app. But for the general case of a MAUI or MAUI Blazor app, you would never dispose MauiApp yourself as that would break your app altogether, and disposing it when the app is closed, to me, doesn't make sense as your app is closing and those resources are going to be released anyway. 

@jonathanpeppers @mattleibow What do ya'll think?

**Comment by malciin** [comment]
@drasticactions I have to say that your answer is pretty... confusing/surprising at least.

> that seems kinda weird to me (Why make the object just to create a new thing off of it and throw it away)

I thought it was _obvious_ this is a minimal reproduction code that does not doing anything meaningful - just reproducing the error. I've added singleton service that needs to be disposed. 

> And if you were doing logic within your app in its Disposed methods to handle when the app is closing... I would rethink that approach.

Really? Disposing object needs rethinking? I can agree that I shouldn't put long running code in disposal but not with that. `Dispose` can do many useful things, some some singleton services could spawn some Process and then in Dispose() could gracefully stopped them. I am little baffled that I need to explain that.

In asp net core or in console app that uses `Microsoft.Extensions.Hosting` everything is correctly disposed at graceful app stop, so I am really surprising that maui blazor framework that not disposing `MauiApp` (that implements `IDisposable`) is a correct approach and I should rethink my app design.

**Comment by drasticactions** [comment]
> @drasticactions I have to say that your answer is pretty... confusing/surprising at least.
> 
> > that seems kinda weird to me (Why make the object just to create a new thing off of it and throw it away)
> 
> I thought it was _obvious_ this is a minimal reproduction code that does not doing anything meaningful - just reproducing the error. I've added singleton service that needs to be disposed and it is not.
> 

I'm sorry! I wasn't talking about your example, I was talking about the MAUI code! 

https://github.com/dotnet/maui/blob/96288c1cff1c6399f42e3e24ea6f5b54557a19c0/src/Core/src/Platform/Windows/MauiWinUIApplication.cs#L32-L36

This specifically creates a MauiApp object, but then doesn't store it anywhere. Even if you wanted to dispose it, you couldn't unless you went out of your way to do so. I wasn't talking about your sample, what you did was 100% fine. I should have been more clear.

**Comment by Eilon** [comment]
Indeed, in normal circumstances app termination is 'violent' - that is, it is both sudden and catastrophic. Very little is cleaned up because the idea is to quickly make the app go away. If anything was truly important, it should have already been done (data was saved, messages sent, etc.).

So why does MauiApp bother with IDisposable? For testing purposes only, really. I mean, you could probably cause it to be disposed with some extra code in your app, so long as you're aware that it still might not get called (process can be easily terminated without _any_ code in the target process being run).

**Comment by Eilon** [comment]
Here's a related PR where I added IAsyncDisposable support, but also only for testing purposes: https://github.com/dotnet/maui/pull/13424

**Comment by Eilon** [comment]
@malciin I think that the scenarios you describe are valid, but I'm not sure they're part of "disposing the MauiApp object". If an app wants to ask the user whether to save/upload/re-combobulate when they try to close the app, then that specific scenario should be handled as part of the "closing the window" event (whatever it might be called).

**Comment by malciin** [comment]
> (..) then that specific scenario should be handled as part of the "closing the window" event (whatever it might be called).

To me thats what `IDisposable` are for. To me thats also what DI simplifies - dispose any created class instances that need disposal. If I register singleton service with IDisposable I, as a programmer, expect that to be disposed. Maybe asking user for save was a bad example, but flushing internal buffers, gracefuly closing tcp connections etc are those examples that I would put in Dispose() and expect them to happend. X should be graceful shutdown, not "violent, sudden and catastrophic".
Many logging libraries writes to files at specific intervals. If MauiApp is not disposed that could ment that some logs could be lost.


Lets put it differently. Why Console app that uses `Microsoft.Extensions.Hosting` can support graceful shutdown, every singleton instance is nicely disposed in that case, buffers are flushed blah blah, but maui **windows app** can not do that? What are differences between them? Why maui app can not dispose itself in main window close event?

**Comment by Eilon** [comment]
I've given this some more thought and it could be worth considering. For anyone interested in this, I have two questions:

1. Under what "app lifecycle event" would you expect the MauiApp to be disposed? (Is there already such a cross-platform MAUI event? Or is there at least one on each platform?)
2. What should we do for IAsyncDisposable? Generally the app lifecycle events are sync only, and there's no safe way to process IAsyncDisposables (we tried something similar in BlazorWebView and it's caused a ton of problems - problems that we don't even know how to fix). Should it do async dispose, but only give it up to let's say 500ms and then terminate? Or some other pattern you can think of?

I think if we can come up with reasonable answers for these then it's worth giving it a shot.

**Comment by ninachen03** [comment]
Verified this issue with Visual Studio 17.10.0 Preview 5 ( 8.0.21 &8.0.0-rc.2.9530).I can repro this issue.

**Comment by NEmbgen** [comment]
> I've given this some more thought and it could be worth considering. For anyone interested in this, I have two questions:
> 
> 1. Under what "app lifecycle event" would you expect the MauiApp to be disposed? (Is there already such a cross-platform MAUI event? Or is there at least one on each platform?)
> 2. What should we do for IAsyncDisposable? Generally the app lifecycle events are sync only, and there's no safe way to process IAsyncDisposables (we tried something similar in BlazorWebView and it's caused a ton of problems - problems that we don't even know how to fix). Should it do async dispose, but only give it up to let's say 500ms and then terminate? Or some other pattern you can think of?
> 
> I think if we can come up with reasonable answers for these then it's worth giving it a shot.

Hi @Eilon,

we are currently running into this with our Multi Window Maui + Blazor App. Blazor components are correctly disposed when a Blazor navigation event happens, but not when we close a window.

On closing a side window very weird things start to happen with our components that **need** to be disposed.

I assume that this is in connection with the blazorwebview that isn't disposed because it still exists (eventhough the window is closed) and it's components try to register themselves again in our service (eventhough they are not displayed in the current webview).


In my ideal world the BlazorWebView an all it's components would be normally disposed (as they alerady are on normal navigation events) when a window is closed via the "x". We do not need async dispose only the "normal" one. The async dispose method could be an iteration on that, once a suitable solution has been found.

I think this would be an imporant task for multi window maui + blazor projects (probably 1 of 5 in the world) :smile:

**Comment by FelixLorenz** [comment]
We experience the "exact same" problem: multi window state gets somewhat out of control because component instances of a closed window still exist.

I wrote about our attempt to get the BlazorWebView disposed as well all "its containing" components in #7277  [here ](https://github.com/dotnet/maui/issues/7277#issuecomment-2143383094).

**Comment by M0n7y5** [comment]
dotnet 9 MAUI + Blazor, same behavior.
